src/asm/charmap.c

```diff
@@ -55,91 +54,104 @@ int32_t charmap_Add(char *input, uint8_t output)
 			charmap = calloc(1, sizeof(struct Charmap));
 			if (charmap == NULL)
 				fatalerror("Not enough memory for charmap");
-
 			pCurrentSection->charmap = charmap;
 		}
 	} else {
 		charmap = &globalCharmap;
 	}
 
-	if (charmap->count > MAXCHARMAPS || strlen(input) > CHARMAPLENGTH)
+	if (charmap->charCount >= MAXCHARMAPS || strlen(input) > CHARMAPLENGTH)
 		return -1;
 
-	input_length = strlen(input);
-	if (input_length > 1) {
-		i = 0;
-		while (i < charmap->count + 1) {
-			if (input_length > strlen(charmap->input[i])) {
-				memcpy(temp1i, charmap->input[i],
-				       CHARMAPLENGTH + 1);
-				memcpy(charmap->input[i], input, input_length);
-				temp1o = charmap->output[i];
-				charmap->output[i] = output;
-				i++;
-				break;
-			}
-			i++;
-		}
-		while (i < charmap->count + 1) {
-			memcpy(temp2i, charmap->input[i], CHARMAPLENGTH + 1);
-			memcpy(charmap->input[i], temp1i, CHARMAPLENGTH + 1);
-			memcpy(temp1i, temp2i, CHARMAPLENGTH + 1);
-			temp2o = charmap->output[i];
-			charmap->output[i] = temp1o;
-			temp1o = temp2o;
-			i++;
+	curr_node = &charmap->nodes[0];
+
+	for (i = 0; (v = (uint8_t)input[i]); i++) {
+		if (curr_node->next[v]) {
+			curr_node = curr_node->next[v];
+		} else {
+			temp_node = &charmap->nodes[charmap->nodeCount + 1];
+
+			curr_node->next[v] = temp_node;
+			curr_node = temp_node;
+
+			++charmap->nodeCount;
 		}
-		memcpy(charmap->input[charmap->count + 1], temp1i,
-		       CHARMAPLENGTH + 1);
-		charmap->output[charmap->count + 1] = temp1o;
-	} else {
-		memcpy(charmap->input[charmap->count], input, input_length);
-		charmap->output[charmap->count] = output;
 	}
-	return ++charmap->count;
+
+	/* prevent duplicated keys by accepting only first key-value pair.  */
+	if (curr_node->isCode)
+		return charmap->charCount;
+
+	curr_node->code 	= output;
+	curr_node->isCode 	= 1;
+
+	return ++charmap->charCount;
 }
 
 int32_t charmap_Convert(char **input)
 {
-	struct Charmap *charmap;
+	struct Charmap 	*charmap;
+	struct Charnode	*charnode;
+
+	char *output;
+	char outchar[8];
 
-	char outchar[CHARMAPLENGTH + 1];
-	char *buffer;
 	int32_t i, j, length;
+	uint8_t v, w;
```

Comment:
The single letter variable names are a bit hard to follow. `i` is OK, since it's just a loop counter, but the `j`, `v`, and `w` variables could have better names. This is just a suggestion.